src/wp-includes/block-template-utils.php

```diff
@@ -519,11 +519,19 @@ function _remove_theme_attribute_from_template_part_block( &$block ) {
  * @return WP_Block_Template Template.
  */
 function _build_block_template_result_from_file( $template_file, $template_type ) {
-	$default_template_types = get_default_block_template_types();
-	$template_content       = file_get_contents( $template_file['path'] );
-	$theme                  = get_stylesheet();
+	$template  = new WP_Block_Template();
+	$cache_key = md5( serialize( $template_file ) );
+	$cache     = wp_cache_get( $cache_key, 'block_template' );
+	if ( is_object( $cache ) ) {
+		foreach ( get_object_vars( $cache ) as $key => $value ) {
+			$template->$key = $value;
+		}
 
-	$template                 = new WP_Block_Template();
+		return $template;
+	}
+
+	$theme                    = ! isset( $template_file['theme'] ) ? $template_file['theme'] : get_stylesheet();
+	$template->content        = file_get_contents( $template_file['path'] );
```

Comment:
@felixarntz The massive performance benefits you saw in your PR are from this line. At the moment, EVERY time a block template is loaded, it is read into memory. In the case of TT4, this means loading, post meta template for every post on the page. In my test site, around 10 times. This is wasteful.